### PR TITLE
Merge analysis flow

### PR DESCRIPTION
**Pull request by andrewkern**
refactor of the snakemake workflow

### CODE REVIEW

File: workflows/inference.snake [hunk 1]
```diff
@@ -47,18 +40,27 @@ if(config["chrm_list"] != "all"):
...
 
 rule all:
     input: 
         #expand(output_dir + "/inference/OutOfAfrica_3G09/polyDFE/{chrms}/pop{ids}.polyDFE.bestfits", chrms=chrm_list, ids=[0, 1, 2])
-        expand(output_dir + "/plots/{demog}/{chrms}/{dfes}/dfe.inference.benchmark.pdf",
+        expand(output_dir + "/plots/{demog}/{chrms}/{dfes}/{annots}/dfe.inference.benchmark.pdf",
             chrms=chrm_list,
-            demog=demo_model_array,
-            dfes=dfe_list)
+            demog=demo_model_ids,
```
[the remark this answers is not in the document]
Comment (xin-huang) [reply workflows/inference.snake:61]: I can edit these parameter names when I update `inference.snake` later.

File: workflows/ts2fs.py [hunk 2]
```diff
@@ -63,7 +63,7 @@ def generate_fs(ts, sample_sets, seq_len, neu_prop, nonneu_prop, output, format)
     nonneu_fs = mut_afs["non_neutral"]
 
     if format == 'dadi': _generate_dadi_fs(neu_fs, nonneu_fs, output)
-    elif format == 'polydfe': _generate_polydfe_fs(neu_fs, nonneu_fs, seq_len, neu_prop, nonneu_prop, output)
+    elif format == 'polydfe': _generate_polydfe_fs(neu_fs, nonneu_fs, seq_len, neu_prop, nonneu_prop, sample_sets, output)
```
Comment (xin-huang) [inline workflows/ts2fs.py:66]: `sample_sets` is wrapped into a list, causing `len(sample_sets) == 1`

Comment (xin-huang) [reply workflows/ts2fs.py:66]: get the `sample_size` before wrapping `sample_sets` into a list (line 61)

Comment (xin-huang) [inline workflows/ts2fs.py:66]:
```suggestion
    elif format == 'polydfe': _generate_polydfe_fs(neu_fs, nonneu_fs, seq_len, neu_prop, nonneu_prop, sample_size, output)
```

File: workflows/ts2fs.py [hunk 3]
```diff
@@ -75,13 +75,13 @@ def _generate_dadi_fs(neu_fs, nonneu_fs, output):
     neu_fs.to_file(output[0])
     nonneu_fs.to_file(output[1])
 
-def _generate_polydfe_fs(neu_fs, nonneu_fs, seq_len, neu_prop, nonneu_prop, output):
+def _generate_polydfe_fs(neu_fs, nonneu_fs, seq_len, neu_prop, nonneu_prop, sample_sets, output):
```
Comment (xin-huang) [inline workflows/ts2fs.py:78]:
```suggestion
def _generate_polydfe_fs(neu_fs, nonneu_fs, seq_len, neu_prop, nonneu_prop, sample_size, output):
```

File: workflows/ts2fs.py [hunk 4]
```diff
@@ -75,13 +75,13 @@ def _generate_dadi_fs(neu_fs, nonneu_fs, output):
...
     """
     """
     neu_len = round(seq_len * neu_prop)
     nonneu_len = round(seq_len * nonneu_prop)
     with open(output[0], 'w') as o:
-        o.write("1 1 20\n")
+        o.write(f"1 1 {len(sample_sets)}\n")
```
Comment (xin-huang) [inline workflows/ts2fs.py:84]:
```suggestion
        o.write(f"1 1 {sample_size}\n")
```

**Review by xin-huang — COMMENTED** [review]
For the DFE part, it works in my machine with slurm.
I have updated `inference.snake`, `ts2fs.py`, and `plots.py` in my repo, I can edit these files after merging this pr.
https://github.com/xin-huang/analysis2/blob/dev/benchmark/inference.snake
https://github.com/xin-huang/analysis2/blob/dev/benchmark/ts2fs.py
https://github.com/xin-huang/analysis2/blob/dev/benchmark/plots.py

File: workflows/ts2fs.py (hunk 2 above)
Comment (andrewkern) [reply workflows/ts2fs.py:66]: so this won't work-- `allele_counts()` needs the list of `sample_sets`. i caught this bug on my version and fixed it however. can you please pull this branch again and see if it gives you the same error?

Comment (xin-huang) [reply workflows/ts2fs.py:66]: Yes, it works after commenting out line 59 and 60, and changing line 45 to `freqs = allele_counts(ts, [sample_sets])`

Comment (andrewkern) [reply workflows/ts2fs.py:66]: okay cool

File: workflows/inference.snake (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewkern) [reply workflows/inference.snake:61]: okay fixed this

File: workflows/n_t.snake [hunk 5]
```diff
@@ -146,43 +136,17 @@ rule download_genetic_map:
...
-            slim_burn_in=10)
-        ts.dump(output[0])
-
-
+def generation_time_helper(wildcards, species):
+    if wildcards.demog == 'Constant':
+        generation_time = species.generation_time
+    else:
+        generation_time = species.get_demographic_model(wildcards.demog).generation_time
+    return generation_time
```
Comment (izabelcavassim) [inline workflows/n_t.snake:149]: When are we simulating the genomes then? In the other workflow?

File: workflows/n_t.snake [hunk 6]
```diff
@@ -202,75 +166,85 @@ rule sp_download:
...
+    )
 
 
 rule compound_stairwayplot:
     input:
-        ne_files
+        rules.run_stairwayplot.output
     output:
-        output_dir + "/stairwayplot_estimated_Ne_{selection}.png"
+        output_dir + "/plots/{demog}/{chrms}/{dfes}/{annots}/dfe.inference.benchmark.pdf/stairwayplot_estimated_Ne.png"
```
Comment (izabelcavassim) [inline workflows/n_t.snake:205]: Is dfe.inference.benchmark.pdf a directory? It looks odd to me.

File: workflows/n_t.snake [hunk 7]
```diff
@@ -283,46 +257,54 @@ rule smcpp_plot:
...
 
 
 rule run_msmc:
     input:
         # TODO make this NOT dependent on all simulations.
         # is there anyway to get access to wildcards.seeds from here? 
-        expand(output_dir + "/Intermediate/{seeds}/{samps}.{chrms}_{selection}.trees.multihep.txt",
-            chrms=chrm_list, seeds=seed_array, samps=num_sampled_genomes_msmc, selection=selection)
-    output: output_dir + "/Intermediate/{seeds}/{samps}.{selection}.trees.multihep.txt.final.txt"
+        #expand(output_dir + "/Intermediate/{seeds}/{samps}.{chrms}_{selection}.trees.multihep.txt",
```
Comment (izabelcavassim) [inline workflows/n_t.snake:292]: I think you can delete if is the same thing as above. I see that above we were previously asking for the mask_file name. Can you include this into the output_directories somehow? Thanks

File: workflows/ts2fs.py [hunk 8]
```diff
@@ -42,7 +42,7 @@ def f(x):
     unique, counts = np.unique(site_class, return_counts=True)
     # print(dict(zip(unique, counts)), flush=True)
 
-    freqs = allele_counts(ts, sample_sets)
+    freqs = allele_counts(ts, [sample_sets])
```
Comment (mufernando) [inline workflows/ts2fs.py:45]: I'm surprised this works?

Comment (mufernando) [inline workflows/ts2fs.py:45]: we should add a note here that this won't work to get joint SFS

File: workflows/n_t.snake [hunk 9]
```diff
@@ -331,18 +313,28 @@ rule run_msmc:
...
+                demog=wildcards.demog,
+                samps=num_sampled_genomes_msmc,
+                annots=wildcards.annots,
+            )
 
 
 rule compound_msmc:
     input: ne_files_msmc,
-    output: output_dir+f"/Results/msmc_estimated_Ne_{selection}.png"
+    output: "/plots/{demog}/{chrms}/{dfes}/{annots}/msmc_estimated_Ne.png"
```
Comment (izabelcavassim) [inline workflows/n_t.snake:337]: Maybe add a directory where mask is included or something that way.

File: workflows/n_t.snake [hunk 10]
```diff
@@ -357,25 +349,29 @@ rule all_plot:
         #f2 = ne_files_smcpp,
         f3 = ne_files_msmc,
     output:
-        output_dir + f"/Results/all_estimated_Ne_{selection}.pdf"
+        output_dir + "/plots/{demog}/{chrms}/{dfes}/{annots}/all_estimated_Ne.pdf"
```
Comment (izabelcavassim) [inline workflows/n_t.snake:352]: cant we have the final pdf with all those names included in the name, such as:
f'all_estimated_Ne_{demog}/{chrms}/{dfes}/{annots}.pdf',
I know it is huge, but I tend to think that this is better than having the same name across different simulation types.

File: workflows/simulation.snake [hunk 11]
```diff
@@ -40,47 +40,67 @@ seed_array = np.random.random_integers(1,2**31,replicates)
 
 # The specific demographic model you would like to run
 demo_model_array =  config["demo_models"]
+demo_sample_size_dict = {}
+for x in demo_model_array:
+    demo_sample_size_dict[x["id"]] = x["num_samples_per_population"]
+demo_model_ids = [x["id"] for x in demo_model_array]
```
Comment (izabelcavassim) [inline workflows/simulation.snake:46]: remember to annotate somewhere that demo_model_ids are a list of strings, and the same for dfe_list and annotation_list. Maybe change the name to 'demo_model_ids_list' to resemble the other two list types?

File: workflows/simulation.snake [hunk 12]
```diff
@@ -40,47 +40,67 @@ seed_array = np.random.random_integers(1,2**31,replicates)
...
-            DFE=dfe,
-        )
+        if wildcards.annots == "all_sites":
+            # Adding selection to the whole contig
+            contig.add_dfe(intervals=np.array([[0, int(contig.length)]]), DFE=dfe)
+        elif wildcards.annots == "none":
+            contig = species.get_contig(wildcards.chrms, genetic_map=genetic_map_id)
+        else:
+          ## Adding annotation only seletion on exon region
+            annot = species.get_annotations(wildcards.annots)
```
Comment (izabelcavassim) [inline workflows/simulation.snake:93]: now what is it happening when we have multiple annotations? We are running one full pipeline for each annotation right? We need to think about how to expand this such that we would want to simulate more than one DFE within the same simulation. But maybe lets wait on that for now.

File: workflows/simulation.snake [hunk 13]
```diff
@@ -40,47 +40,67 @@ seed_array = np.random.random_integers(1,2**31,replicates)
...
+        
         contig.mutation_rate = mutation_rate
         engine = stdpopsim.get_engine("slim")
         ts = engine.simulate(
             model,
             contig,
             samples,
-            slim_scaling_factor=10,
+            slim_scaling_factor=config["slim_scaling_factor"],
             slim_burn_in=10,
```
Comment (izabelcavassim) [inline workflows/simulation.snake:104]: the slim_burn_in is fine at 10?

File: workflows/stairway.py [hunk 14]
```diff
@@ -44,7 +44,7 @@ def __init__(self, workdir, stairway_dir, java_exe="java"):
         self.classpath = "{}:{}".format(stairway_path, stairway_path / "swarmops.jar")
         self.java_exe = java_exe
 
-    def ts_to_stairway(self, ts_path, selection, num_bootstraps=1, mask_file=None):
+    def ts_to_stairway(self, ts_path, num_bootstraps=1, mask_file=None):
```
Comment (izabelcavassim) [inline workflows/stairway.py:47]: I need to update this part by extracting the nonsynonymous part of the SFS. I will open a PR for that once this one is in.

**Review by izabelcavassim — COMMENTED** [review]
OK, I gave my 2 cents worth review ;). 
I think we can delete the folder n_with_s as well right @andrewkern?
I also think that we can improve and/or add a README of the whole pipeline, so anyone that is jumping in just now would be able to follow it. 

Thanks @andrewkern

File: workflows/n_t.snake (hunk 10 above)
Comment (andrewkern) [reply workflows/n_t.snake:352]: right now everything is sorted by directory. we could change this later if needed

File: workflows/simulation.snake (hunk 13 above)
Comment (andrewkern) [reply workflows/simulation.snake:104]: i'll change this to be read from the config file

File: workflows/msmc.py [hunk 15]
```diff
@@ -50,12 +50,9 @@ def write_msmc_file(path, num_sampled_genomes_msmc, selection, mask_file=None):
         dirr = os.path.dirname(path)
         filen = os.path.basename(path)
         sep = filen.split(".")
-        print(sep)
         chrom = sep[0].split("_")[0]
-        print("this is chrom")
-        print(chrom)
         sep.insert(0, str(sample_size))
-        output = os.path.join(dirr, ".".join(sep) + f".multihep.txt")
+        #output = os.path.join(dirr, ".".join(sep) + ".multihep.txt")
```
[the remark this answers is not in the document]
Comment (andrewkern) [reply workflows/msmc.py:55]: `output` is a parameter of `write_msmc_file` -- no reason to mess with it here

File: workflows/n_t.snake [hunk 16]
```diff
@@ -31,19 +31,23 @@ import plots
...
 # for each replicate
 # TODO double check this is up to date with stdpopsim backend 
 num_samples_per_population = config["num_samples_per_population"]
 population_id = config["population_id"]
+
 # Here is a list of sample sizes to run msmc on. 
 # Each element counts as its own analysis
 # so there will be "replicates" runs for each size
-num_sampled_genomes_msmc = [int(x) for x in config["num_sampled_genomes_msmc"].split(",")]
+num_sampled_genomes_msmc =  config["num_sampled_genomes_msmc"]
```
[the remark this answers is not in the document]
Comment (andrewkern) [reply workflows/n_t.snake:50]: yes we need to redo the documentation here. will do that in a separate PR

**Pull request merged**